packages/common/src/type-helpers.ts

```diff
@@ -73,3 +86,45 @@ export function errorCode(error: unknown): string | undefined {
 export function assertNever(msg: string, x: never): never {
   throw new TypeError(msg + ': ' + x);
 }
+
+/**
+ * According to the EcmaScript's spec, the default behavior of JavaScript's `x instanceof Y` operator is to walk up the
+ * prototype chain of object 'x', checking if any constructor in that hierarchy is _exactly the same object_ as the
+ * constructor function 'Y'.
+ *
+ * Unfortunately, it happens in various situations that different constructor function objects get created for what
+ * appears to be the very same class. This leads to surprising behavior where `instanceof` returns false though it is
+ * known that the object is indeed an instance of that class. One particular case where this happens is when constructor
+ * 'Y' belongs to a different realm than the constuctor with which 'x' was instantiated. Another case is when two copies
+ * of the same library gets loaded in the same realm.
+ *
+ * In practice, this tends to cause issues when crossing the workflow-sandboxing boundary (since Node's vm module
+ * really creates new execution realms), as well as when running tests using Jest (see https://github.com/jestjs/jest/issues/2549
+ * for some details on that one).
+ *
+ * This function injects a custom 'instanceof' handler into the prototype of 'clazz', which is both cross-realm safe and
+ * cross-copies-of-the-same-lib safe. It works by adding a special symbol property to the prototype of 'clazz', and then
+ * checking for the presence of that symbol.
+ */
+export function makeCrossRealmSafeInstanceOf(clazz: { prototype: any }, markerName: string): void {
+  const marker = Symbol.for(`__temporal_is${markerName}`);
+
+  Object.defineProperty(clazz.prototype, marker, { value: true, enumerable: false });
+  Object.defineProperty(clazz, Symbol.hasInstance, {
+    // eslint-disable-next-line object-shorthand
+    value: function (this: any, error: object): boolean {
+      if (this === clazz) {
+        return isRecord(error) && (error as any)[marker] === true;
+      } else {
+        // 'this' must be a _subclass_ of clazz that doesn't redefined [Symbol.hasInstance], so that it inherited
+        // from clazz's [Symbol.hasInstance]. If we don't handle this particular situation, then
+        // `x instanceof SubclassOfParent` would return true for any instance of 'Parent', which is clearly wrong.
+        //
+        // Ideally, it'd be preferable to avoid this case entirely, by making sure that all subclasses of 'clazz'
+        // redefine [Symbol.hasInstance], but we can't enforce that. We therefore fallback to the default instanceof
+        // behavior (which is NOT cross-realm safe).
+        return this.prototype.isPrototypeOf(error); // eslint-disable-line no-prototype-builtins
+      }
```

Comment:
If that was only about our own code, I'd agree.

However, if users themselves subclass one of our class (some type of specific `ApplicationFailure` for example, though we strongly discourage doing that), then I do not expect them to make it cross realm safe. The helper function is not even exposed publicly, and I don't think we should expect users to understand low details of these issues. In that case, making `instanceof` fails systematically would likely cause more damage than it would resolve.

An alternative I have been considering, however, would be to use a simple constructor’s name comparision in those cases. This would save users from a low probability of false negative, but introduces an even lower probability of false positive. In the end, I think however it might turns out being more predictable for users.

Reply:
It makes sense... And for our code, is there are a way we could make it more robust, e.g., throwing in debug mode during testing, so that we can make sure we never miss a subclass...

Reply:
I'm running out of time, so I'll leave this the way it is for now, but I was indeed thinking about introducing some "paranoid validation" to be enabled during tests to assert that we are not missing any context-related issues. That would probably not be in this `else` case however (or _not only_ in there), as it would be very likely that we miss some other objects. Might try hook [around calls to the sandbox](https://github.com/mjameswh/sdk-typescript/blob/aa737e22937c3b258391a47c8a848404c422957b/packages/worker/src/workflow/vm.ts#L47-L61).